### ISSUE TITLE
Document the existance of obj.getBrailleRegions(review=review)

### ISSUE THREAD

**Issue by derekriemer** [nvaccess/nvda#7125]
Hi:
This method is not present in the base NVDAObject, thus it's not apparent in the API Docs for an NVDAObject this can be overridden. Can this be documented somehow?

**Comment by derekriemer** [comment]
@jcsteh What if we put this in raising "notImplementedError" and did the optimization we do for globalPlugins with chooseNVDAObjectOverlayClasses?

**Comment by jcsteh** [comment]
I think that'd probably be premature optimisation; we could always do that if we found it to be a notable performance problem. Really, it's just a matter of adding it and documenting it.

**Comment by derekriemer** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [Jun 28, 2017, 8:16 PM MDT](https://github.com/nvaccess/nvda/issues/7125#issuecomment-311843221 "2017-06-29T02:16:00Z - Replied by Github Reply Comments"):
> I think that'd probably be premature optimisation; we could always do that if we found it to be a notable performance problem. Really, it's just a matter of adding it and documenting it.

It's not optimization, it's a matter of the code calling it catching an attribute error and if it exists, simply not continuing through the function anymore. (I think)? Is this not intended behavior? I can remove the try/accept in this code, and kill the return if need be.
Also, we could put that code in the base for getBrailleRegions

```
def getFocusRegions(obj, review=False):
	# Allow objects to override normal behaviour.
	try:
		regions = obj.getBrailleRegions(review=review)
	except (AttributeError, NotImplementedError):
		pass
	else:
		for region in regions:
			region.update()
			yield region
		return

	# Late import to avoid circular import.
	from treeInterceptorHandler import TreeInterceptor, DocumentTreeInterceptor
	from cursorManager import CursorManager
	from NVDAObjects import NVDAObject
	if isinstance(obj, CursorManager):
		region2 = (ReviewTextInfoRegion if review else CursorManagerRegion)(obj)
	elif isinstance(obj, DocumentTreeInterceptor) or (isinstance(obj,NVDAObject) and NVDAObjectHasUsefulText(obj)): 
		region2 = (ReviewTextInfoRegion if review else TextInfoRegion)(obj)
	else:
		region2 = None
	if isinstance(obj, TreeInterceptor):
		obj = obj.rootNVDAObject
	region = NVDAObjectRegion(obj, appendText=TEXT_SEPARATOR if region2 else "")
	region.update()
	yield region
	if region2:
		region2.update()
		yield region2


```

**Comment by jcsteh** [comment]
[**@derekriemer**](https://github.com/derekriemer) commented on [1 Jul 2017, 19:31 GMT+10](https://github.com/nvaccess/nvda/issues/7125#issuecomment-312421717 "2017-07-01T09:31:50Z - Replied by Github Reply Comments"):
> It's not optimization, it's a matter of the code calling it catching an attribute error and if it exists, simply not continuing through the function anymore.

Checking whether it's in the dict would be an optimisation (probably a premature one). Once this is added to the base NVDAObject, the AttributeError catch should be removed.

> Also, we could put that code in the base for getBrailleRegions

Not sure if I'm perhaps misunderstanding you, but we don't want the base implementation of getBrailleRegions to do anything; it should just raise NotImplementedError. The reason is that NVDAObjects should try to avoid containing base presentation logic wherever possible. I realise this isn't always the case - there are cases where we do this where we shouldn't and cases where it just makes sense - but this is presentation logic that is used in 99% of cases. Overriding the default braille regions is something that should almost always be avoided, which is one of the reasons I was reluctant to document getBrailleRegions or have a base method in the first place. However, it's the only way to do something like math interaction, so I accept there are use cases where it's necessary.

**Comment by bhavyashah** [comment]
@derekriemer Is the argument/explanation provided by @jcsteh in https://github.com/nvaccess/nvda/issues/7125#issuecomment-312426558 sufficient for you? If not, I would encourage further discussion so as to ascertain what needs to be done.

**Comment by derekriemer** [comment]
yeah, just haven't had time for a PR yet, as I had a computer crash and work.

**Comment by leonardder** [comment]
Personally, I vote for adding this on the base object, raising NotImplemented. This will also allow people to use super when overriding this method, e.g.

```
def getBrailleRegions(review=False):
    if something:
        yield someRegion()
        return
    super(Class,self).getBrailleRegions(review=review)
```

**Comment by Adriani90** [comment]
cc: @josephsl

**Comment by Adriani90** [comment]
This needs probably to be revisited now under Python 3.11.
cc: @SaschaCowley, @michaelDCurran
